pkg/operator/controller/canary/daemonset_test.go

```diff
@@ -56,4 +58,71 @@ func TestDesiredCanaryDaemonSet(t *testing.T) {
 	if !cmp.Equal(containers[0].Image, canaryImage) {
 		t.Errorf("expected daemonset container image to be %q, but got %q", canaryImage, containers[0].Image)
 	}
+
+	nodeSelector := daemonset.Spec.Template.Spec.NodeSelector
+	expectedNodeSelector := map[string]string{
+		"kubernetes.io/os": "linux",
+	}
+	if !cmp.Equal(nodeSelector, expectedNodeSelector) {
+		t.Errorf("expected daemonset node selector to be %q, but got %q", expectedNodeSelector, nodeSelector)
+
+	}
+}
+
+func TestCanaryDaemonsetChanged(t *testing.T) {
+	testCases := []struct {
+		description string
+		mutate      func(*appsv1.DaemonSet)
+		expect      bool
+	}{
+		{
+			description: "if nothing changes",
+			mutate:      func(_ *appsv1.DaemonSet) {},
+			expect:      false,
+		},
+		{
+			description: "if pod template node selector changes",
+			mutate: func(ds *appsv1.DaemonSet) {
+				ds.Spec.Template.Spec.NodeSelector = map[string]string{
+					"foo":  "bar",
+					"test": "selector",
+				}
+			},
+			expect: true,
+		},
+		{
+			description: "if pod template tolerations changes",
+			mutate: func(ds *appsv1.DaemonSet) {
+				ds.Spec.Template.Spec.Tolerations = []corev1.Toleration{
+					{
+						Key:      "test",
```

Comment:
Maybe add a check the functionality of the comparator, as this just checks for the existence of a new toleration.

Reply:
`cmpTolerations` is for sorting a slice of tolerations. This test does exercise a comparison between 2 different tolerations as is.

Reply:
Ive added a quick tolerations check to `TestDesiredCanaryDaemonSet` to highlight this (and since that coverage was missing).

Reply:
Can one field of a toleration change to indicate a daemonset change?  I.e. if I change the Value or Operator of a toleration, do you mark the daemonset changed?

Reply:
Changing any part of the toleration slice would trigger an update, yea. Should we add more test cases here? At a certain point we would just be testing that `cmp.Equal` works as expected.

Reply:
It's worth noting that `cmpTolerations` is copied from cluster-ingress-operator, which already has unit tests.  In particular, cluster-ingress-operator has a test case where the key changes:  https://github.com/openshift/cluster-ingress-operator/blob/7b41dc34fedc5107a8fe79a2a16d718b070aea09/pkg/operator/controller/ingress/deployment_test.go#L636-L643

It might be worth moving some of this common code to a shared package at some point.